src/bin/muxer/avformat.rs

```diff
@@ -0,0 +1,123 @@
+// PATENTS file, you can obtain it at www.aomedia.org/license/patent.
+
+#[cfg(feature = "ffmpeg-sys")]
+use ffmpeg_sys::*;
+
+#[allow(unused_imports)]
+use super::Muxer;
+#[allow(unused_imports)]
+use std::ffi::CString;
+#[allow(unused_imports)]
+use std::io;
+#[allow(unused_imports)]
+use std::mem;
+#[allow(unused_imports)]
+use std::ptr;
+
+#[cfg(feature = "ffmpeg-sys")]
+pub struct AvformatMuxer {
+  context: *mut AVFormatContext,
+  duration: i64
+}
+
+#[cfg(feature = "ffmpeg-sys")]
+impl AvformatMuxer {
+  pub fn open(path: &str) -> Box<dyn Muxer> {
+    unsafe {
+      let mut context = ptr::null_mut();
+      let p = CString::new(path).expect("convert to cstring failed");
+      match avformat_alloc_output_context2(
+        &mut context,
+        ptr::null_mut(),
+        ptr::null(),
+        p.as_ptr() as *const i8
+      ) {
+        0 => {
+          let mut io = ptr::null_mut();
+          let ret =
+            avio_open(&mut io, p.as_ptr() as *const i8, AVIO_FLAG_WRITE);
+          if ret < 0 {
+            panic!("open ouput file {} failed", path);
+          }
+          (*context).pb = io;
+          Box::new(AvformatMuxer { context, duration: 0 })
+        }
+        e => panic!("open ouput failed, error = {}", e)
+      }
+    }
+  }
+}
+
+#[cfg(feature = "ffmpeg-sys")]
+impl Muxer for AvformatMuxer {
+  fn write_header(
+    &mut self, width: usize, height: usize, framerate_num: usize,
+    framerate_den: usize
+  ) {
+    unsafe {
+      let stream = avformat_new_stream(self.context, ptr::null_mut());
+      if stream.is_null() {
+        panic!("new stream failed");
+      }
+      let param = (*stream).codecpar;
+      (*param).codec_type = AVMediaType::AVMEDIA_TYPE_VIDEO;
+      (*param).codec_id = AVCodecID::AV_CODEC_ID_AV1;
+      (*param).width = width as i32;
+      (*param).height = height as i32;
+      let ret = avformat_write_header(self.context, ptr::null_mut());
+      if ret < 0 {
+        panic!("write header failed error = {}", ret);
+      }
+      let framerate_den = framerate_den as i32;
+      let framerate_num = framerate_num as i32;
+      let duration = (*stream).time_base.den * framerate_den
+        / ((*stream).time_base.num * framerate_num);
+      self.duration = duration as i64;
+    }
+  }
+
+  fn write_frame(&mut self, pts: u64, data: &[u8]) {
+    unsafe {
+      let mut pkt: AVPacket = mem::zeroed();
+      av_init_packet(&mut pkt);
+      pkt.data = data.as_ptr() as *mut _;
+      pkt.size = data.len() as i32;
+
+      let pts = pts as i64;
+      pkt.pts = pts * self.duration;
+      //no b frame, pts equals dts
+      pkt.dts = pts * self.duration;
+      pkt.duration = self.duration;
+      let ret = av_write_frame(self.context, &mut pkt);
+      if ret < 0 {
+        panic!("write frame failed error = {}", ret);
+      }
+    }
+  }
+
+  fn flush(&mut self) -> io::Result<()> {
+    unsafe {
+      let io = (*self.context).pb;
+      avio_flush(io);
```

Comment:
Shouldn't flushing also entail calling `av_write_frame` until it returns 1?

Reply:
yes, your are right.

Reply:
Just check more code. Actually flush is not so useful for container format.
If we did not call av_write_trailer. the entire stream is not parseable even we flush all things.
Anyway, I will change the code to follow av_write_frame document.